### ISSUE TITLE
plugin: add test (was encoding/asn1 error when using plugin)

### ISSUE THREAD

**Issue by fsenart** [golang/go#18584]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?

`go version go1.8beta2 linux/amd64`

### What operating system and processor architecture are you using (`go env`)?

```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/fsenart/projects"
GORACE=""
GOROOT="/home/fsenart/tools/go"
GOTOOLDIR="/home/fsenart/tools/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build746500890=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
```

### What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
A link on play.golang.org is best.

```go
// main.go

package main

import (
	"plugin"
	"reflect"
)

func main() {
	p, err := plugin.Open("plugin.so")
	if err != nil {
		panic(err)
	}
	f, err := p.Lookup("F")
	if err != nil {
		panic(err)
	}

	// f.(func())()
	reflect.ValueOf(f).Call(nil)
}
```

```go
// plugin.go

package main

import "C"
import "net/http"

func F() {
	_, err := http.Get("https://example.com")
	if err != nil {
		panic(err)
	}
}
```

```Makefile
# Makefile

test: build
	@./main

build: clean
	@go build -buildmode=plugin -o plugin.so plugin.go
	@go build -o main main.go

clean:
	@rm -rf main *.so
```

### What did you expect to see?

I don't expect any panic to happen. 

### What did you see instead?

I see a panic happen.

```
panic: Get https://example.com: tls: failed to parse certificate from server: asn1: structure error: tags don't match (4 vs {class:0 tag:16 length:1242 isCompound:true}) {optional:false explicit:false application:false defaultValue:<nil> tag:<nil> stringType:0 timeType:0 set:false omitEmpty:false} RawContent @4

goroutine 1 [running]:
plugin/unnamed-4ad3198f73971e7223d350ce7ae956cd80d817d9.F()
	/home/fsenart/projects/src/github.com/alsanium/issue_asn1/plugin.go:9 +0x65
reflect.Value.call(0x4d3820, 0xc42000f728, 0x13, 0x4ef099, 0x4, 0x0, 0x0, 0x0, 0xc42000c2e0, 0x4d3820, ...)
	/home/fsenart/tools/go/src/reflect/value.go:434 +0x91f
reflect.Value.Call(0x4d3820, 0xc42000f728, 0x13, 0x0, 0x0, 0x0, 0xc42004df68, 0x4be551, 0x0)
	/home/fsenart/tools/go/src/reflect/value.go:302 +0xa4
main.main()
	/home/fsenart/projects/src/github.com/alsanium/issue_asn1/main.go:19 +0xed
Makefile:2: recipe for target 'test' failed
```

---

### Observations

* If I don't use the `reflect` package, and instead make type assertion manually, then no error happens.
  ```go
  f.(func())()
  // reflect.ValueOf(f).Call(nil)
  ```

* If I don't use the `plugin` package, then no error happens.

  ```go
  package main

  import (
      "net/http"
      "reflect"
  )

  var f interface{} = func() {
      _, err := http.Get("https://example.com")
      if err != nil {
          panic(err)
      }
  }

  func main() {
      //f.(func())()
      reflect.ValueOf(f).Call(nil)
  }

  ```

**Comment by bradfitz** [comment]
Probably too late for a Go 1.8 fix, but maybe a point release.

**Comment by fsenart** [comment]
OK. THIS IS WEIRD.

After more investigation, it seems not even related to the `reflect` package. 

The error happens if you import *any* package when using the `plugin` package.
For example (`plugin.go` as above):

```go
// main.go

package main

import "plugin"

// you can import nearly any package (tested with fmt, log, json, etc.)
import "encoding/json"
var dummy json.RawMessage

func main() {
	p, err := plugin.Open("plugin.so")
	if err != nil {
		panic(err)
	}
	f, err := p.Lookup("F")
	if err != nil {
		panic(err)
	}

	f.(func())()
}
```

**Comment by stefanahman** [comment]
Getting the same error using https://github.com/eawsy/aws-lambda-go-shim library.
After more investigation and reading this thread, I noticed the `plugin` import in this [file](https://github.com/eawsy/aws-lambda-go-shim/blob/6fb5482c187723979d8aee72b5e8e7b011978025/src/shim.go#L35).

**Comment by fsenart** [comment]
@stefanahman we effectively use go plugins in this project and the issue is related to the fact that we have other imports than `plugin` in the referenced file (cf. [above comment](#issuecomment-271454470)).

@bradfitz at the time I've opened the issue, I agree with you that the problem was very special and happen in rare cases and Go1.8Maybe milestone was ok. 
But considering my [last comment](#issuecomment-271454470), you must notice that as soon as any package is imported along with the plugin package then **all SSL requests fail**. Isn't it a major issue to be fixed before the release of 1.8? Can you please reconsider the milestone, taking into account the new elements?

**Comment by crawshaw** [comment]
Importing package `plugin` into a binary switches the linker to the dynamic linking codepath. Presumably that is triggering something which causes type-handling problems.

(I won't have access to a linux machine to investigate this until tomorrow.)

**Comment by cristim** [comment]
@fsenart, @bradfitz IMHO having the new plugin feature break TLS when there are any other imports makes it pretty useless for a lot of web backend use cases, so I think 1.8 shouldn't be released before this issue is fixed.

**Comment by bradfitz** [comment]
If we stopped the release train for every bug, we'd never ship.

Plugins are a new part of Go 1.8 anyway. Expect some rough edges. For instance, we disabled them entirely on Mac at the last second, and they don't work on Windows. A few more bugs are expected. That's what point releases are for.

**Comment by cristim** [comment]
@bradfitz if you mentioned all that it sounds reasonable, I was just under the impression that the plugin feature was a bit more mature than that.

**Comment by schickling** [comment]
Same issue here. Thanks for investigating @fsenart 👍

**Comment by fsenart** [comment]
The issue (or at least its visible symptoms) is resolved in go1.8rc1. /cc @crawshaw @bradfitz 
PS: I let you close the issue or remain it open for deep investigations.

**Comment by bradfitz** [comment]
@crawshaw, @ianlancetaylor do you know what might've fixed this between beta2 and rc1?

**Comment by ianlancetaylor** [comment]
It was fixed by the fix to #18252.

But we should add a test for this in misc/cgo/testplugin.

**Comment by stefanahman** [comment]
Can confirm that my issue has been resolved with `v1.8rc1` 😃 🎉
